### PR TITLE
PROP-27 - Implement Registry proxy

### LINKED ISSUE #27: Feature: Implement Registry proxy
### Is your feature request related to a problem? Please describe.

No

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Implement proxy service to OCI registry.

It will be MQTT-to-HTTP proxy, as RP will receive commands via MQTT channels and then fetch Wasm modules from OCI registry via HTTP, and then send them to the Proplet in chunks via MQTT channel.

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nyagamunene**
# What type of PR is this?

## Which issue(s) does this PR fix/relate to?
- Resolves #27

### CODE REVIEW

File: cmd/proxy/.env [hunk 1]
```diff
@@ -0,0 +1,9 @@
+MQTT_BROKER_URL=mqtt://localhost:1883
+MQTT_PASSWORD=0e8d1d8d-e3b8-4c20-8873-df200fb56100
```
Comment (rodneyosodo) [inline cmd/proxy/.env:2]: Rename

```
MQTT_REGISTRY_CHANNEL_ID=
MQTT_REGISTRY_THING_ID=
MQTT_REGISTRY_THING_KEY=
```

File: cmd/proxy/main.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+const (
+	mqttPrefix = "MQTT_"
+	httpPrefix = "HTTP_"
+	chanSize   = 2
+)
+
+func main() {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:25]:
```suggestion
	ctx, cancel := context.WithCancel(context.Background())
```

File: cmd/proxy/main.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	defer cancel()
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	logger := slog.New(slog.NewTextHandler(os.Stdout, nil))
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:31]: Use same as manager

File: cmd/proxy/main.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	err := godotenv.Load()
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:33]: Do we need this?

File: cmd/proxy/main.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	cfgM, err := config.LoadMQTTConfig(env.Options{Prefix: mqttPrefix})
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:38]: mqttConfig

File: proxy/config/http.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+type HTTPProxyConfig struct {
+	RegistryURL  string `env:"REGISTRY_URL" envDefault:"localhost:5000"`
+	Authenticate bool   `env:"AUTHENTICATE" envDefault:"false"`
+	Username     string `env:"USERNAME"     envDefault:""`
+	Password     string `env:"PASSWORD"     envDefault:""`
+}
+
+func LoadHTTPConfig(opts env.Options) (*HTTPProxyConfig, error) {
+	config := HTTPProxyConfig{}
```
Comment (rodneyosodo) [inline proxy/config/http.go:24]: move this to main. environment variables should be loaded from main

File: proxy/mqtt/mqtt.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	token := c.client.Subscribe(subTopic, 1, handler)
+	if err := token.Error(); err != nil {
+		return fmt.Errorf("failed to subscribe to %s: %w", subTopic, err)
+	}
+
+	return nil
+}
+
+func (c *RegistryClient) PublishContainer(ctx context.Context, containerData []byte) error {
+	pubTopic := fmt.Sprintf("channels/%s/messages/registry/server", c.config.ChannelID)
```
Comment (rodneyosodo) [inline proxy/mqtt/mqtt.go:104]: define these topics at the top then use variable subsitution when using them

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: proxy/config/http.go (hunk 6 above)
Comment (nyagamunene) [reply proxy/config/http.go:24]: This is called in the main file but its implemented in the config file. So should I remove implementation from the config file?

File: cmd/proxy/main.go (hunk 3 above)
Comment (rodneyosodo) [reply cmd/proxy/main.go:25]: This has not been resolved

File: cmd/proxy/main.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	httpPrefix = "HTTP_"
+)
+
+func main() {
+	g, ctx := errgroup.WithContext(context.Background())
+
+	logger := slog.New(slog.NewTextHandler(os.Stdout, nil))
+	slog.SetDefault(logger)
+
+	err := godotenv.Load("cmd/proxy/.env")
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:28]: No need for this. Use what is similar to magistrala

File: cmd/proxy/main.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+	logger.Info("starting proxy service")
+
+	if err := start(ctx, g, service); err != nil {
+		logger.Error(fmt.Sprintf("%s service exited with error: %s", svcName, err))
+	}
+}
+
+func start(ctx context.Context, g *errgroup.Group, s *proxy.ProxyService) error {
+	slog.Info("connecting...")
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:66]: No need for this

File: proxy/config/http.go (hunk 6 above)
Comment (rodneyosodo) [reply proxy/config/http.go:34]: Yes. Move it to main. Loading env variables should be done on main

File: proxy/config/mqtt.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+type MQTTProxyConfig struct {
+	BrokerURL string `env:"BROKER_URL" envDefault:""`
+	Password  string `env:"PASSWORD"   envDefault:""`
+	PropletID string `env:"PROPLET_ID" envDefault:""`
+	ChannelID string `env:"CHANNEL_ID" envDefault:""`
+}
+
+func LoadMQTTConfig(opts env.Options) (*MQTTProxyConfig, error) {
+	c := MQTTProxyConfig{}
```
Comment (rodneyosodo) [inline proxy/config/mqtt.go:15]: same

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: cmd/proxy/main.go (hunk 3 above)
Comment (rodneyosodo) [reply cmd/proxy/main.go:21]:
```go
var level slog.Level
if err := level.UnmarshalText([]byte(cfg.LogLevel)); err != nil {
	return fmt.Errorf("failed to parse log level: %s", err.Error())
}
logHandler := slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{
	Level: level,
})
logger := slog.New(logHandler)
slog.SetDefault(logger)
```

File: proxy/README.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
+# Proxy Service
+
+The Proxy Service acts as a bridge between MQTT and HTTP protocols in the Propeller system. It enables bidirectional communication between MQTT clients and HTTP endpoints, allowing for seamless integration of different protocols.
```
Comment (rodneyosodo) [inline proxy/README.md:3]: Move this to docs repo

File: proxy/config/http.go [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+	RegistryURL  string `env:"PROXY_REGISTRY_URL"      envDefault:""`
+	ChunkSize    int    `env:"PROXY_CHUNK_SIZE"        envDefault:"512000"`
+	Authenticate bool   `env:"PROXY_AUTHENTICATE"      envDefault:"false"`
+	Token        string `env:"PROXY_REGISTRY_TOKEN"    envDefault:""`
+	Username     string `env:"PROXY_REGISTRY_USERNAME" envDefault:""`
+	Password     string `env:"PROXY_REGISTRY_PASSWORD" envDefault:""`
+}
+
+func (c *HTTPProxyConfig) Validate() error {
+	if c.RegistryURL == "" {
```
Comment (rodneyosodo) [inline proxy/config/http.go:34]: This validation can be done on the env using https://github.com/caarlos0/env

File: proxy/config/mqtt.go [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+type MQTTProxyConfig struct {
+	BrokerURL string `env:"PROXY_MQTT_ADDRESS" envDefault:"tcp://localhost:1883"`
+	Password  string `env:"PROXY_PROPLET_KEY"  envDefault:""`
+	PropletID string `env:"PROXY_PROPLET_ID"   envDefault:""`
+	ChannelID string `env:"PROXY_CHANNEL_ID"   envDefault:""`
+}
+
+func (c *MQTTProxyConfig) Validate() error {
+	if c.BrokerURL == "" {
```
Comment (rodneyosodo) [inline proxy/config/mqtt.go:17]: same use https://github.com/caarlos0/env to validate

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: cmd/proxy/main.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+	"github.com/absmach/propeller/proxy"
+	"github.com/absmach/propeller/proxy/config"
+	"github.com/caarlos0/env/v11"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	svcName  = "proxy"
+	logLevel = "info"
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:18]: This should be an env variable

File: cmd/proxy/main.go [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+	if err := level.UnmarshalText([]byte(logLevel)); err != nil {
+		log.Fatalf("failed to parse log level: %s", err.Error())
+	}
+	logHandler := slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{
+		Level: level,
+	})
+	logger := slog.New(logHandler)
+	slog.SetDefault(logger)
+
+	mqttCfg := config.MQTTProxyConfig{}
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:34]: There is some stutter here in config MQTT Proxy Config

You can move the configs to the proxy package and import it as 

```go
mqttConfig := proxy.MQTTConfig{}
```

File: proplet/requests.go [hunk 16]
```diff
@@ -3,10 +3,11 @@ package proplet
...
 
 type startRequest struct {
-	ID           string
-	FunctionName string
-	WasmFile     []byte
-	Params       []uint64
+	ID                   string
+	FunctionName         string
+	WasmFile             []byte
+	WasmFileDownloadPath string
```
Comment (rodneyosodo) [inline proplet/requests.go:9]: Can this be changed to image URL?

File: proxy/service.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+	containerChunks := make(map[string]int)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case chunk := <-s.dataChan:
+			if err := s.mqttClient.PublishContainer(ctx, chunk); err != nil {
+				s.logger.Error("failed to publish container chunk",
+					"error", err,
```
Comment (rodneyosodo) [inline proxy/service.go:87]: use logger args

File: proxy/service.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+					"chunk", chunk.ChunkIdx,
+					"total", chunk.TotalChunks)
+
+				continue
+			}
+
+			s.logger.Info("published container chunk",
+				"chunk_name", chunk.AppName,
+				"chunk_no", chunk.ChunkIdx,
```
Comment (rodneyosodo) [inline proxy/service.go:96]: same

File: proxy/service.go [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+			if err := s.mqttClient.PublishContainer(ctx, chunk); err != nil {
+				s.logger.Error("failed to publish container chunk",
+					"error", err,
+					"chunk", chunk.ChunkIdx,
+					"total", chunk.TotalChunks)
+
+				continue
+			}
+
+			s.logger.Info("published container chunk",
```
Comment (rodneyosodo) [inline proxy/service.go:94]: Should this be debug information?

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: test.md [hunk 20]
```diff
@@ -102,3 +102,28 @@ export PROPLET_THING_ID=""
 export PROPLET_THING_KEY=""
 propeller-proplet
 ```
+
```
Comment (rodneyosodo) [inline test.md:105]: Move this to https://github.com/absmach/propeller-docs/pull/6 and delete it here

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: cmd/proxy/main.go [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+	LogLevel string `env:"PROXY_LOG_LEVEL"           envDefault:"info"`
+
+	BrokerURL  string `env:"PROXY_MQTT_ADDRESS"          envDefault:"tcp://localhost:1883"`
+	PropletKey string `env:"PROXY_PROPLET_KEY,notEmpty"`
+	PropletID  string `env:"PROXY_PROPLET_ID,notEmpty" `
+	ChannelID  string `env:"PROXY_CHANNEL_ID,notEmpty"`
+
+	ChunkSize    int    `env:"PROXY_CHUNK_SIZE"             envDefault:"512000"`
+	Authenticate bool   `env:"PROXY_AUTHENTICATE"           envDefault:"false"`
+	Token        string `env:"PROXY_REGISTRY_TOKEN"         envDefault:""`
```
Comment (rodneyosodo) [inline cmd/proxy/main.go:27]: Do we need token and password or it is either one?

Comment (nyagamunene) [reply cmd/proxy/main.go:27]: Its either of the 2

**Review by rodneyosodo — APPROVED** [review]

File: Makefile [hunk 22]
```diff
@@ -22,9 +22,9 @@ $(SERVICES):
 
 install:
 	for file in $(BUILD_DIR)/*; do \
-		if [[ ! "$$file" =~ \.wasm$$ ]]; then \
-			cp "$$file" $(GOBIN)/propeller-`basename "$$file"`; \
-		fi \
+		if [ "$$file" = "$${file%%.wasm}" ]; then \
```
Comment (drasko) [inline Makefile:25]: Can we use Makefile `wildcard` directive here to contract the filtering and selection for copying?

**Review by drasko — CHANGES_REQUESTED** [review]

File: Makefile [hunk 23]
```diff
@@ -20,12 +20,10 @@ endef
...
 
+
+# Install all non-WASM executables from the build directory to GOBIN with 'propeller-' prefix
 install:
-	for file in $(BUILD_DIR)/*; do \
-		if [[ ! "$$file" =~ \.wasm$$ ]]; then \
-			cp "$$file" $(GOBIN)/propeller-`basename "$$file"`; \
-		fi \
-	done
+	$(foreach f,$(wildcard $(BUILD_DIR)/*[!.wasm]),cp $(f) $(patsubst $(BUILD_DIR)/%,$(GOBIN)/propeller-%,$(f));)
```
Comment (drasko) [inline Makefile:26]: Would it be possible, in order to simplify handling of binaries, to have a subdir `wasm` within `build` dir, and put all Wasm binaries there. I guess these binaries are coming from examples, right? If really needed you can also add `build/bin` and `build/wasm`, but I do not think that we need `build/bin` (as most of the time we'll need just Propeller services (and not examples), and they can live directly in the top `build` dir, for convenience), but if this simplifies copying, then please go ahead and make both `build/bin` and `build/wasm`.

**Review by drasko — CHANGES_REQUESTED** [review]

File: Makefile (hunk 23 above)
Comment (nyagamunene) [reply Makefile:26]: > Would it be possible, in order to simplify handling of binaries, to have a subdir `wasm` within `build` dir, and put all Wasm binaries there. I guess these binaries are coming from examples, right? If really needed you can also add `build/bin` and `build/wasm`, but I do not think that we need `build/bin` (as most of the time we'll need just Propeller services (and not examples), and they can live directly in the top `build` dir, for convenience), but if this simplifies copying, then please go ahead and make both `build/bin` and `build/wasm`.

We can implement this .

**Pull request merged**